src/nspawn/nspawn.c

```diff
@@ -2725,6 +2753,33 @@ static int inner_child(
         return log_error_errno(r, "execv() failed: %m");
 }
 
+static int setup_sd_notify_child(void) {
+        static const int one = 1;
+        int fd;
+        union sockaddr_union sa = {
+                .sa.sa_family = AF_UNIX,
+        };
+        int r;
+
+        fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0)
+                return log_error_errno(errno, "Failed to allocate notification socket: %m");
+
+        (void) mkdir_parents_label(nspawn_notify_socket_path, 0755);
+        (void) unlink(nspawn_notify_socket_path);
+
+        strncpy(sa.un.sun_path, nspawn_notify_socket_path, sizeof(sa.un.sun_path)-1);
+        r = bind(fd, &sa.sa, offsetof(struct sockaddr_un, sun_path) + strlen(sa.un.sun_path));
+        if (r < 0)
+                return log_error_errno(errno, "bind(%s) failed: %m", sa.un.sun_path);
```

Comment:
usually we write something like this then:
```c
r = fd;
fd = -1;
return r;
```

or so. That said, some times it is indeed nicer to just explicitly call safe_close() in the failure if block instead (and not use _cleanup_close_), if you only have one of those, for example.